File: python/main/edrm/EDRMBuilder.py
from pathlib import Path
from typing import Any
from copy import deepcopy

from xml.dom.minidom import getDOMImplementation, Element, Document

from edrm.DirectoryEntry import DirectoryEntry
from edrm.EntryInterface import EntryInterface
from edrm.FileEntry import FileEntry
from edrm.MappingEntry import MappingEntry
import edrm


class EDRMBuilder:
    def __init__(self):
        self.__output_path: Path = Path("./output.xml").resolve().absolute()
        self.__as_nli: bool = False

        self.__entries: dict[str, EntryInterface] = {}
        self.__families: dict[str, list[str]] = {}

    @property
    def output_path(self) -> Path:
        return self.__output_path

    @output_path.setter
    def output_path(self, output_path: Path) -> None:
        self.__output_path = output_path

    @property
    def as_nli(self) -> bool:
        return self.__as_nli

    @as_nli.setter
    def as_nli(self, as_nli: bool) -> None:
        self.__as_nli = as_nli

    @property
    def entry_map(self) -> dict[str, EntryInterface]:
        """
        Retrieve a dicstionary mapping each EntryInterface to its id.  Note: the returned dictionary cannot be used to
        modify the builder's collection of entries.
        :return: dictionary of strings representing an entry's id to the corresponding EntryInterface object
        """
        return deepcopy(self.__entries)

    def add_entry(self, entry: EntryInterface) -> str:
        entry_id = entry[entry.identifier_field].value
        self.__entries[entry_id] = entry
        parent_id = entry.parent

        if entry_id not in self.__families.keys():
            self.__families[entry_id] = []

        if parent_id is not None:
            family = self.__families.get(parent_id, [])
            family.append(entry_id)
            self.__families[parent_id] = family

        return entry_id

    def add_file(self, file_path: str, mimetype: str, parent_id: str = None) -> str:
        return self.add_entry(FileEntry(file_path, mimetype, parent_id))

    def add_directory(self, directory: str, parent_id: str = None) -> str:
        return self.add_entry(DirectoryEntry(directory, parent_id))

    def add_mapping(self, mapping: dict[str, Any], mimetype: str, parent_id: str = None) -> str:
        return self.add_entry(MappingEntry(mapping, mimetype, parent_id))

    def __add_folder(self, doc_id: str, document: Document, container: Element, families):
        doc_element = document.createElement('Document')
        doc_element.setAttribute('DocId', doc_id)
        container.appendChild(doc_element)

        if doc_id not in families:
            return

        family = families.pop(doc_id)
        if len(family) > 0:
            folder_element = document.createElement('Folder')
            folder_element.setAttribute('FolderName', doc_id)
            container.appendChild(folder_element)

            for child in self.__families[doc_id]:
                self.__add_folder(child, document, folder_element, families)

    def build(self) -> Document:
        impl = getDOMImplementation()
        doc = impl.createDocument(None, 'Root', None)
        root = doc.documentElement

        root.setAttribute('MajorVersion', '1')
        root.setAttribute('MinorVersion', '2')
        root.setAttribute('Description', 'EDRM XML Load File')
        root.setAttribute('Locale', 'US')
        root.setAttribute('DataInterchangeType', 'Update')

        field_list = doc.createElement("Fields")
        root.appendChild(field_list)
        for entry_id, entry in self.__entries.items():
            entry.serialize_field_definitions(doc, field_list)

        batch_element = doc.createElement('Batch')
        root.appendChild(batch_element)
        doc_list = doc.createElement('Documents')
        batch_element.appendChild(doc_list)

        for entry_id, entry in self.__entries.items():
            entry.serialize_entry(doc, doc_list, self.__entries, self.as_nli)

        relationship_list = doc.createElement('Relationships')
        batch_element.appendChild(relationship_list)
        for parent_id, family in self.__families.items():
            for child_id in family:
                relationship = doc.createElement('Relationship')
                relationship.setAttribute('Type', 'Container')
                relationship.setAttribute('ParentDocId', parent_id)
                relationship.setAttribute('ChildDocId', child_id)
                relationship_list.appendChild(relationship)

        folders_list = doc.createElement('Folders')
        batch_element.appendChild(folders_list)

        working_families = deepcopy(self.__families)
        for entry_id in self.__families.keys():
            if entry_id in working_families:
                self.__add_folder(entry_id, doc, folders_list, working_families)

        return doc

    def save(self, doc: Document = None):
        if doc is None:
            doc = self.build()

        with self.output_path.open(mode='w', encoding=edrm.configs['encoding']) as load_file:
            doc.writexml(load_file, encoding=edrm.configs['encoding'], standalone=True, addindent='  ', newl='\n')

File: java/src/main/java/com/nuix/nli/edrm/EDRMBuilder.java
package com.nuix.nli.edrm;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class EDRMBuilder {
    private final Map<String, EntryInterface> entryMap = new LinkedHashMap<>();
    private final Map<String, List<String>> familyMap = new LinkedHashMap<>();
    private final EntryInterface EMPTY = new MappingEntry(Map.of(), "application/x-empty");

    private boolean asNli = false;
    private Path outputPath = Path.of(".", "output.xml").normalize().toAbsolutePath();

    public boolean isAsNli() { return asNli; }
    public void setAsNli(boolean asNli) { this.asNli = asNli; }

    public Path getOutputPath() { return outputPath; }
    public void setOutputPath(Path outputPath) { this.outputPath = outputPath; }

    public Map<String, EntryInterface> getEntryMap() { return new LinkedHashMap<>(entryMap); }

    public String addEntry(EntryInterface entry) {
        String id = entry.getField(entry.getIdentifierField()).getValue().toString();
        entryMap.put(id, entry);

        String parentId = entry.getParentId();
        if (!familyMap.containsKey(id))
            familyMap.put(id, new LinkedList<String>());

        if (null != parentId && !parentId.isEmpty()) {
            List<String> family = familyMap.getOrDefault(parentId, new LinkedList<>());
            family.add(id);
            familyMap.put(parentId, family);
        }

        return id;
    }

    public EntryInterface getEntry(String id) {
        if (null == id || !entryMap.containsKey(id)) return EMPTY;
        return entryMap.get(id);
    }

    public String addFile(String filePath, String mimetype) { return addFile(filePath, mimetype, null); }
    public String addFile(String filePath, String mimetype, String parentId) {
        return addEntry(new FileEntry(filePath, mimetype, parentId));
    }

    public String addDirectory(String directory) { return addDirectory(directory, null); }
    public String addDirectory(String directory, String parentId) {
        return addEntry(new DirectoryEntry(directory, parentId));
    }

    public String addMapping(Map<String, Object> mapping, String mimetype) { return addMapping(mapping, mimetype, null); }
    public String addMapping(Map<String, Object> mapping, String mimetype, String parentId) {
        return addEntry(new MappingEntry(mapping, mimetype, parentId));
    }

    private void addFolder(String docId, Document document, Element container, Map<String, List<String>> familyMap) {
        EntryInterface docEntry = getEntry(docId);
        if (null != docEntry.getParentId()) {
            Element docElement = document.createElement("Document");
            docElement.setAttribute("DocId", docId);
            container.appendChild(docElement);
        }

        if (!familyMap.containsKey(docId)) {
            return;
        }

        List<String> family = familyMap.remove(docId);
        if(!family.isEmpty()) {
            Element folderElement = document.createElement("Folder");
            folderElement.setAttribute("DocId", docId);
            container.appendChild(folderElement);
            for (String childId : this.familyMap.get(docId)) {
                addFolder(childId, document, folderElement, familyMap);
            }
        }
    }

    public Document build() {
        try {
            Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
            Element root = doc.createElement("Root");
            doc.appendChild(root);
            root.setAttribute("MajorVersion", "1");
            root.setAttribute("MinorVersion", "2");
            root.setAttribute("Description", "EDRM XML Load File");
            root.setAttribute("Locale", "US");
            root.setAttribute("DataInterchangeType", "Update");

            Element fields = doc.createElement("Fields");
            root.appendChild(fields);

            for (EntryInterface e : entryMap.values()) {
                e.serializeFieldDefinitions(doc, fields);
            }

            Element batch = doc.createElement("Batch");
            root.appendChild(batch);
            Element documentsList = doc.createElement("Documents");
            batch.appendChild(documentsList);

            for (Map.Entry<String, EntryInterface> entry : entryMap.entrySet()) {
                entry.getValue().serializeEntry(doc, documentsList, entryMap, asNli);
            }

            Element relationships = doc.createElement("Relationships");
            batch.appendChild(relationships);
            for (Map.Entry<String, List<String>> entry : familyMap.entrySet()) {
                String parentId = entry.getKey();
                List<String> family = entry.getValue();

                for (String childId : family) {
                    Element relationship = doc.createElement("Relationship");
                    relationship.setAttribute("ParentDocId", parentId);
                    relationship.setAttribute("ChildDocId", childId);
                    relationship.setAttribute("RelationshipType", "Container");
                    relationships.appendChild(relationship);
                }
            }

            Element foldersList = doc.createElement("Folders");
            batch.appendChild(foldersList);

            LinkedHashMap<String, List<String>> familyMapCopy = new LinkedHashMap<>(familyMap);
            for (String entryID : familyMap.keySet()) {
                if (familyMapCopy.containsKey(entryID)) {
                    addFolder(entryID, doc, foldersList, familyMapCopy);
                }
            }

            return doc;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public void save(Document doc) {
        if (null == outputPath) throw new IllegalStateException("outputPath not set");

        if (null == doc) {
            doc = build();
        }

        try {
            Files.createDirectories(outputPath.getParent());
            Transformer t = TransformerFactory.newInstance().newTransformer();
            t.setOutputProperty(OutputKeys.INDENT, "yes");
            t.setOutputProperty(OutputKeys.METHOD, "xml");
            t.setOutputProperty(OutputKeys.ENCODING, EDRMUtilities.EDRM_CONFIG.getOrDefault("encoding", "UTF-8"));
            t.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
            t.transform(new DOMSource(doc), new StreamResult(outputPath.toFile()));
        } catch (IOException | javax.xml.transform.TransformerException e) {
            throw new RuntimeException(e);
        }
    }

    public void save() {
        save(null);
    }
}